Create Demo/Test App for OpenAPI Java Client

Create a demo/test application which uses the OpenAPI Java client.

@Jeeppler 
I try to generate openapi3.json using
`./gradlew generateOpenapi`
but get a lot the same testing errors:
> java.lang.IllegalStateException: Could not initialize plugin: interface org.mockito.plugins.MockMaker
> Could not initialize inline Byte Buddy mock maker.
> It appears as if your JDK does not supply a working agent attachment mechanism.
> Java               : 11
> JVM vendor name    : Debian
> JVM vendor version : 11.0.16+8-post-Debian-1deb11u1
> JVM name           : OpenJDK 64-Bit Server VM
> JVM version        : 11.0.16+8-post-Debian-1deb11u1
> JVM info           : mixed mode, sharing
> OS name            : Linux
> OS version         : 5.10.102.1-microsoft-standard-WSL2
> Caused by: java.lang.IllegalStateException: Could not self-attach to current VM using external process

It seems to me the problem is in my virtual machine. I'm using Debian on the Windows 10 host. I would like to ask what environment do you use for the Sechub building?

@Jeeppler 
Maybe I found a little bug. 
When I build Java client library with

> ./gradlew :sechub-api-java:buildAPIJava

I see this error

> /mnt/sechub/sechub-api-java/gen/src/main/java/com/mercedesbenz/sechub/api/java/model/ProjectWhiteList.java:41: error: class ProjectWhitelist is public, should be declared in a file named ProjectWhitelist.java
> public class ProjectWhitelist implements Serializable {

It seems to me that there is a misspelling in the generated class name. It should be 
`public class ProjectWhiteList implements Serializable {`
but not
`public class ProjectWhitelist implements Serializable {`

--- a/sechub-api-java/build.gradle
+++ b/sechub-api-java/build.gradle
@@ -6,6 +6,13 @@
  * ============================================================================
  */
 
+ext {
+    jersey_version = "2.35"
+    jackson_version = "2.13.2"
+    jackson_databind_version = "2.13.2.2"
+    jackson_databind_nullable_version = "0.2.3"
+}
+
 dependencies {
     // The production code uses the SLF4J logging API at compile time
     implementation spring_boot_dependency.slf4j_api
@@ -17,11 +24,24 @@ dependencies {
     implementation library.apache_httpcomponents_client
 
     testImplementation spring_boot_dependency.junit_jupiter
-    
-    
+
     testImplementation spring_boot_dependency.mockito_core
     testImplementation library.mockito_inline
     testImplementation spring_boot_dependency.hamcrest
+
+    implementation library.jcommander
+    implementation spring_boot_dependency.logback_classic
+    implementation "org.glassfish.jersey.core:jersey-client:$jersey_version"
+    implementation "org.glassfish.jersey.inject:jersey-hk2:$jersey_version"
+    implementation "org.glassfish.jersey.media:jersey-media-multipart:$jersey_version"
+    implementation "org.glassfish.jersey.media:jersey-media-json-jackson:$jersey_version"
+    implementation "org.glassfish.jersey.connectors:jersey-apache-connector:$jersey_version"
+    implementation "com.fasterxml.jackson.core:jackson-core:$jackson_version"
+    implementation "com.fasterxml.jackson.core:jackson-annotations:$jackson_version"
+    implementation "com.fasterxml.jackson.core:jackson-databind:$jackson_databind_version"
+    implementation "com.fasterxml.jackson.datatype:jackson-datatype-jsr310:$jackson_version"
+    implementation "org.openapitools:jackson-databind-nullable:$jackson_databind_nullable_version"
+    implementation files('gen/build/libs/sechub-api-java-0.0.0.jar')
 }
 
 // see https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator-gradle-plugin/README.adoc
@@ -115,3 +135,18 @@ task generateAPIJava(dependsOn: callOpenAPIJavaGenerator){
     }
 }
 
+task buildOpenAPITestTool(type: Jar, dependsOn: build) {
+    group 'sechub'
+    description 'Builds the SecHub Open API CLI test tool.'
+    archiveBaseName = 'sechub-OpenAPI-TestTool'
+
+    manifest {
+        attributes 'Main-Class': 'com.mercedesbenz.sechub.api.java.demo.cli.OpenAPITestTool'
+    }
+
+    from {
+        configurations.runtimeClasspath.collect { it.isDirectory() ? it : zipTree(it) }
+    }
+    duplicatesStrategy = DuplicatesStrategy.INCLUDE
+    with jar
+}

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/SecHubClient.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/SecHubClient.java
@@ -4,11 +4,8 @@ package com.mercedesbenz.sechub.api.java;
 import java.io.File;
 import java.io.IOException;
 import java.net.URI;
-import java.net.URISyntaxException;
 
 import javax.crypto.SealedObject;
-
-import org.apache.http.client.utils.URIBuilder;
 
 import com.fasterxml.jackson.core.JsonParseException;
 import com.fasterxml.jackson.databind.DeserializationFeature;
@@ -28,31 +25,22 @@ public class SecHubClient {
     private String username;
     private SealedObject sealedApiToken;
     private URI hostUri;
-    private int hostPort;
     private boolean trustAll;
 
     private CryptoAccess<String> apiTokenAccess = new CryptoAccess<>();
 
-    public static SecHubClient create(String username, String apiToken, String hostUri, int hostPort) {
-        return create(username, apiToken, hostUri, hostPort, false);
+    public static SecHubClient create(String username, String apiToken, URI hostUri) {
+        return create(username, apiToken, hostUri, false);
     }
 
-    public static SecHubClient create(String username, String apiToken, String hostUri, int hostPort, boolean trustAll) {
-        URI baseUri = null;
-
-        try {
-            baseUri = new URIBuilder(hostUri).build();
-        } catch (URISyntaxException e) {
-            throw new IllegalArgumentException("Not an acceptable uri:" + hostUri, e);
-        }
-        return new SecHubClient(username, apiToken, baseUri, hostPort, trustAll);
+    public static SecHubClient create(String username, String apiToken, URI hostUri, boolean trustAll) {
+        return new SecHubClient(username, apiToken, hostUri, trustAll);
     }
 
-    private SecHubClient(String username, String apiToken, URI hostUri, int hostPort, boolean trustAll) {
+    private SecHubClient(String username, String apiToken, URI hostUri, boolean trustAll) {
         this.username = username;
         this.sealedApiToken = apiTokenAccess.seal(apiToken);
         this.hostUri = hostUri;
-        this.hostPort = hostPort;
         this.trustAll = trustAll;
     }
 
@@ -66,10 +54,6 @@ public class SecHubClient {
 
     public URI getHostUri() {
         return hostUri;
-    }
-
-    public int getHostPort() {
-        return hostPort;
     }
 
     public boolean isTrustAll() {

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/CommandLineParser.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/CommandLineParser.java
@@ -1,0 +1,51 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.api.java.demo.cli;
+
+import com.beust.jcommander.JCommander;
+import com.beust.jcommander.ParameterException;
+import com.mercedesbenz.sechub.api.java.SecHubClient;
+import com.mercedesbenz.sechub.api.java.demo.cli.exceptions.SecHubClientConfigurationRuntimeException;
+
+public class CommandLineParser {
+
+    private JCommander commander;
+
+    /**
+     * Parses given arguments
+     *
+     * @param args
+     * @return configuration or <code>null</code> when only help wanted
+     * @throws SecHubClientConfigurationRuntimeException
+     */
+    public SecHubClient parse(String... args) throws SecHubClientConfigurationRuntimeException {
+        CommandLineSettings settings = parseCommandLineParameters(args);
+        if (settings.isHelpRequired()) {
+            showHelp();
+            return null;
+        }
+        SecHubClientConfigurationFactory configFactory = new SecHubClientConfigurationFactory();
+        return configFactory.create(settings);
+    }
+
+    private CommandLineSettings parseCommandLineParameters(String... args) throws SecHubClientConfigurationRuntimeException {
+        CommandLineSettings settings = new CommandLineSettings();
+        /* @formatter:off */
+        commander = JCommander.newBuilder()
+                                .programName("OpenAPITestTool")
+                                .addObject(settings)
+                                .acceptUnknownOptions(false)
+                              .build();
+        /* @formatter:on */
+        try {
+            commander.parse(args);
+            return settings;
+        } catch (ParameterException e) {
+            throw new SecHubClientConfigurationRuntimeException("Parsing command line parameters failed!", e);
+        }
+    }
+
+    private void showHelp() {
+        commander.usage();
+    }
+
+}

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/CommandLineSettings.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/CommandLineSettings.java
@@ -11,20 +11,12 @@ public class CommandLineSettings {
         return help;
     }
 
-    @Parameter(names = { "--serverUri" }, description = "Specifies the Sechub Server URI. You can also set the environment variable "
-            + EnvironmentVariableConstants.SECHUB_SERVER_URI + ", instead of using this parameter.", required = false)
-    private String serverUri;
+    @Parameter(names = { "--server" }, description = "Specifies the Sechub Server URI. You can also set the environment variable "
+            + EnvironmentVariableConstants.SECHUB_SERVER + ", instead of using this parameter.", required = false)
+    private String server;
 
-    public String getServerUri() {
-        return serverUri;
-    }
-
-    @Parameter(names = { "--serverPort" }, description = "Specifies the Sechub Server port. You can also set the environment variable "
-            + EnvironmentVariableConstants.SECHUB_SERVER_PORT + ", instead of using this parameter.", required = false)
-    private int serverPort;
-
-    public int getServerPort() {
-        return serverPort;
+    public String getServer() {
+        return server;
     }
 
     @Parameter(names = { "--userId" }, description = "Specifies the Sechub Server privileged user id. You can also set the environment variable "
@@ -41,6 +33,15 @@ public class CommandLineSettings {
 
     public String getApiToken() {
         return apiToken;
+    }
+
+    @Parameter(names = {
+            "--trustAll" }, description = "When set to true, then HTTPS certificate checking will be disabled. May be useful when using self-signed certificates. Please try to avoid this setting for security reasons. You can also set the environment variable "
+                    + EnvironmentVariableConstants.SECHUB_TRUSTALL + ", instead of using this parameter.", required = false)
+    private String trustAll;
+
+    public String getTrustAll() {
+        return trustAll;
     }
 
 }

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/CommandLineSettings.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/CommandLineSettings.java
@@ -1,0 +1,46 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.api.java.demo.cli;
+
+import com.beust.jcommander.Parameter;
+
+public class CommandLineSettings {
+    @Parameter(names = { "--help" }, description = "Shows help and provides information on how to use OpenAPI Test Tool.", help = true)
+    private boolean help;
+
+    public boolean isHelpRequired() {
+        return help;
+    }
+
+    @Parameter(names = { "--serverUri" }, description = "Specifies the Sechub Server URI. You can also set the environment variable "
+            + EnvironmentVariableConstants.SECHUB_SERVER_URI + ", instead of using this parameter.", required = false)
+    private String serverUri;
+
+    public String getServerUri() {
+        return serverUri;
+    }
+
+    @Parameter(names = { "--serverPort" }, description = "Specifies the Sechub Server port. You can also set the environment variable "
+            + EnvironmentVariableConstants.SECHUB_SERVER_PORT + ", instead of using this parameter.", required = false)
+    private int serverPort;
+
+    public int getServerPort() {
+        return serverPort;
+    }
+
+    @Parameter(names = { "--userId" }, description = "Specifies the Sechub Server privileged user id. You can also set the environment variable "
+            + EnvironmentVariableConstants.SECHUB_USERID + ", instead of using this parameter.", required = false)
+    private String userId;
+
+    public String getUserId() {
+        return userId;
+    }
+
+    @Parameter(names = { "--apiToken" }, description = "Specifies the privileged user's api token. You can also set the environment variable "
+            + EnvironmentVariableConstants.SECHUB_APITOKEN + ", instead of using this parameter.", required = false)
+    private String apiToken;
+
+    public String getApiToken() {
+        return apiToken;
+    }
+
+}

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/EnvironmentVariableConstants.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/EnvironmentVariableConstants.java
@@ -1,0 +1,14 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.api.java.demo.cli;
+
+public class EnvironmentVariableConstants {
+
+    public static final String SECHUB_SERVER_URI = "SECHUB_SERVER_URI";
+
+    public static final String SECHUB_SERVER_PORT = "SECHUB_SERVER_PORT";
+
+    public static final String SECHUB_USERID = "SECHUB_USERID";
+
+    public static final String SECHUB_APITOKEN = "SECHUB_APITOKEN";
+
+}

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/EnvironmentVariableConstants.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/EnvironmentVariableConstants.java
@@ -3,12 +3,12 @@ package com.mercedesbenz.sechub.api.java.demo.cli;
 
 public class EnvironmentVariableConstants {
 
-    public static final String SECHUB_SERVER_URI = "SECHUB_SERVER_URI";
-
-    public static final String SECHUB_SERVER_PORT = "SECHUB_SERVER_PORT";
+    public static final String SECHUB_SERVER = "SECHUB_SERVER";
 
     public static final String SECHUB_USERID = "SECHUB_USERID";
 
     public static final String SECHUB_APITOKEN = "SECHUB_APITOKEN";
+
+    public static final String SECHUB_TRUSTALL = "SECHUB_TRUSTALL";
 
 }

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/EnvironmentVariableReader.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/EnvironmentVariableReader.java
@@ -14,28 +14,4 @@ public class EnvironmentVariableReader {
     public String readAsString(String environmentVariable) {
         return System.getenv(environmentVariable);
     }
-
-    /**
-     * Reads the value of the environment variable {@link environmentVariable} and
-     * returns the value as integer.
-     *
-     * @param environmentVariable
-     * @return Value of environmentVariable as integer or 0 if the environment
-     *         variable is not set
-     * @throws IllegalArgumentException If the value of the environment variable
-     *                                  {@link environmentVariable} is no valid
-     *                                  integer.
-     */
-    public int readAsInt(String environmentVariable) {
-        try {
-            String value = System.getenv(environmentVariable);
-            if (value == null) {
-                return 0;
-            }
-            return Integer.parseInt(value);
-        } catch (NumberFormatException e) {
-            throw new IllegalArgumentException("The value of " + environmentVariable + " is not a valid integer.");
-        }
-    }
-
 }

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/EnvironmentVariableReader.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/EnvironmentVariableReader.java
@@ -1,0 +1,41 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.api.java.demo.cli;
+
+public class EnvironmentVariableReader {
+
+    /**
+     * Reads the value of the environment variable {@link environmentVariable} and
+     * returns the value as string.
+     *
+     * @param environmentVariable
+     * @return Value of environmentVariable as string or null if the environment
+     *         variable is not set
+     */
+    public String readAsString(String environmentVariable) {
+        return System.getenv(environmentVariable);
+    }
+
+    /**
+     * Reads the value of the environment variable {@link environmentVariable} and
+     * returns the value as integer.
+     *
+     * @param environmentVariable
+     * @return Value of environmentVariable as integer or 0 if the environment
+     *         variable is not set
+     * @throws IllegalArgumentException If the value of the environment variable
+     *                                  {@link environmentVariable} is no valid
+     *                                  integer.
+     */
+    public int readAsInt(String environmentVariable) {
+        try {
+            String value = System.getenv(environmentVariable);
+            if (value == null) {
+                return 0;
+            }
+            return Integer.parseInt(value);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("The value of " + environmentVariable + " is not a valid integer.");
+        }
+    }
+
+}

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/OpenAPITestTool.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/OpenAPITestTool.java
@@ -39,9 +39,9 @@ public class OpenAPITestTool {
             secHubClient = parser.parse(args);
             LOG.info("Granted the following Sechub connection parameters:");
             LOG.info("*** Sechub server URI: " + secHubClient.getHostUri());
-            LOG.info("*** Sechub server port: " + secHubClient.getHostPort());
             LOG.info("*** Privileged user id: " + secHubClient.getUsername());
             LOG.info("*** Privileged user's API token: " + secHubClient.getSealedApiToken());
+            LOG.info("*** trustAll: " + secHubClient.isTrustAll());
         } catch (SecHubClientConfigurationRuntimeException e) {
             LOG.error("An error occurred while parsing the command line arguments: ", e);
             throw new OpenAPITestToolRuntimeException("Connection configuration was invalid.");

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/OpenAPITestTool.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/OpenAPITestTool.java
@@ -1,0 +1,51 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.api.java.demo.cli;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.mercedesbenz.sechub.api.java.SecHubClient;
+import com.mercedesbenz.sechub.api.java.demo.cli.exceptions.OpenAPITestToolRuntimeException;
+import com.mercedesbenz.sechub.api.java.demo.cli.exceptions.SecHubClientConfigurationRuntimeException;
+
+public class OpenAPITestTool {
+    private static final Logger LOG = LoggerFactory.getLogger(OpenAPITestTool.class);
+
+    SecHubClient secHubClient = null;
+
+    public static void main(String[] args) {
+        new OpenAPITestTool().start(args);
+    }
+
+    private void start(String[] args) {
+        try {
+            LOG.info("Building connection configuration.");
+            secHubClient = resolveConfiguration(args);
+
+            LOG.info("Testing Sechub server using OpenAPI library.");
+            SechubServerTest sechubServerTest = new SechubServerTest(secHubClient);
+
+            sechubServerTest.runAllTests();
+            LOG.info("Sechub server successfully tested.");
+        } catch (OpenAPITestToolRuntimeException e) {
+            LOG.error("Sechub server testing failed: ", e);
+        }
+
+    }
+
+    private SecHubClient resolveConfiguration(String[] args) {
+        CommandLineParser parser = new CommandLineParser();
+        try {
+            secHubClient = parser.parse(args);
+            LOG.info("Granted the following Sechub connection parameters:");
+            LOG.info("*** Sechub server URI: " + secHubClient.getHostUri());
+            LOG.info("*** Sechub server port: " + secHubClient.getHostPort());
+            LOG.info("*** Privileged user id: " + secHubClient.getUsername());
+            LOG.info("*** Privileged user's API token: " + secHubClient.getSealedApiToken());
+        } catch (SecHubClientConfigurationRuntimeException e) {
+            LOG.error("An error occurred while parsing the command line arguments: ", e);
+            throw new OpenAPITestToolRuntimeException("Connection configuration was invalid.");
+        }
+        return secHubClient;
+    }
+}

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/README.adoc
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/README.adoc
@@ -44,7 +44,17 @@ Open API library must be in the following folder:
 gen/build/libs/sechub-api-java-0.0.0.jar
 ----
 
-When you want to build the OpenAPI Test Tool, just do:
+When you want to build the OpenAPI Test Tool, do next:
+
+- generate Open API json file
+----
+sechub$ ./gradlew generateOpenapi -x:sechub-api-java:compileJava
+----
+- build Open API library
+----
+sechub$ ./gradlew :sechub-api-java:buildAPIJava
+----
+- build OpenAPI Test Tool
 ----
 sechub$ ./gradlew buildOpenAPITestTool
 ----

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/README.adoc
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/README.adoc
@@ -1,0 +1,50 @@
+// SPDX-License-Identifier: MIT
+== OpenAPI Test Tool
+
+CLI Test Tool for OpenAPI
+
+=== Usage
+
+This is a small CLI tool created for testing SecHub OpenApi library.
+It must be configured by using the command line parameters or the environment variables. All configurations in a way that the command line parameters always overwrite the environment variables. *All parameters are required.*
+
+Ennvironment variables:
+----
+SECHUB_SERVER_URI=https://localhost
+SECHUB_SERVER_PORT=8443
+SECHUB_USERID=int-test_superadmin
+SECHUB_APITOKEN=int-test_superadmin-pwd
+----
+
+Command line paremeters:
+----
+Usage: OpenAPITestTool [options]
+  Options:
+    --apiToken
+      Specifies the privileged user's api token. You can also set the
+      environment variable SECHUB_APITOKEN, instead of using this parameter.
+    --help
+      Shows help and provides information on how to use OpenAPI Test Tool.
+    --serverPort
+      Specifies the Sechub Server port. You can also set the environment
+      variable SECHUB_SERVER_PORT, instead of using this parameter.
+      Default: 0
+    --serverUri
+      Specifies the Sechub Server URI. You can also set the environment
+      variable SECHUB_SERVER_URI, instead of using this parameter.
+    --userId
+      Specifies the Sechub Server privileged user id. You can also set the
+      environment variable SECHUB_USERID, instead of using this parameter.
+----
+
+=== Build
+
+Open API library must be in the following folder:
+----
+gen/build/libs/sechub-api-java-0.0.0.jar
+----
+
+When you want to build the OpenAPI Test Tool, just do:
+----
+sechub$ ./gradlew buildOpenAPITestTool
+----

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SecHubClientConfigurationFactory.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SecHubClientConfigurationFactory.java
@@ -16,19 +16,16 @@ public class SecHubClientConfigurationFactory {
 
     public SecHubClient create(CommandLineSettings settings) {
         URI serverUri = null;
-        if (settings.getServerUri() != null)
-            serverUri = URI.create(settings.getServerUri());
+        if (settings.getServer() != null)
+            serverUri = URI.create(settings.getServer());
 
-        int serverPort = settings.getServerPort();
         String userId = settings.getUserId();
         String apiToken = settings.getApiToken();
+        String trustAll = settings.getTrustAll();
 
-        String envServerUri = environmentVariableReader.readAsString(EnvironmentVariableConstants.SECHUB_SERVER_URI);
+        String envServerUri = environmentVariableReader.readAsString(EnvironmentVariableConstants.SECHUB_SERVER);
         if (serverUri == null && envServerUri != null) {
             serverUri = URI.create(envServerUri);
-        }
-        if (serverPort <= 0 || serverPort > 65535) {
-            serverPort = environmentVariableReader.readAsInt(EnvironmentVariableConstants.SECHUB_SERVER_PORT);
         }
         if (userId == null) {
             userId = environmentVariableReader.readAsString(EnvironmentVariableConstants.SECHUB_USERID);
@@ -36,15 +33,13 @@ public class SecHubClientConfigurationFactory {
         if (apiToken == null) {
             apiToken = environmentVariableReader.readAsString(EnvironmentVariableConstants.SECHUB_APITOKEN);
         }
+        if (trustAll == null) {
+            trustAll = environmentVariableReader.readAsString(EnvironmentVariableConstants.SECHUB_TRUSTALL);
+        }
 
         if (!isServerUriValid(serverUri)) {
             throw new SecHubClientConfigurationRuntimeException(
-                    "Sechub server's URI is invalid. Please set the Sechub server's URI as an environment variable (SECHUB_SERVER_URI) or as a program argument (--serverUri).");
-        }
-
-        if (serverPort <= 0 || serverPort > 65535) {
-            throw new SecHubClientConfigurationRuntimeException("Sechub server's port is set to " + serverPort
-                    + ". Please set the Sechub server's port as an environment variable (SECHUB_SERVER_PORT) or as a program argument (--serverPort).");
+                    "Sechub server's URI is invalid. Please set the Sechub server's URI as an environment variable (SECHUB_SERVER) or as a program argument (--server).");
         }
 
         if (userId == null) {
@@ -57,18 +52,29 @@ public class SecHubClientConfigurationFactory {
                     "Privileged user's API token is null. Please set the privileged user's API token as an environment variable (SECHUB_APITOKEN) or as a program argument (--apiToken).");
         }
 
-        return SecHubClient.create(userId, apiToken, serverUri.toString(), serverPort);
+        if (!isTrustAllValid(trustAll)) {
+            throw new SecHubClientConfigurationRuntimeException(
+                    "trustAll value is invalid. Please set the trustAll as an environment variable (SECHUB_TRUSTALL) or as a program argument (--trustAll). Correct values are true and false.");
+        }
+        return trustAll == null ? SecHubClient.create(userId, apiToken, serverUri)
+                : SecHubClient.create(userId, apiToken, serverUri, Boolean.parseBoolean(trustAll));
     }
 
     private boolean isServerUriValid(URI serverUri) {
         if (serverUri == null)
             return false;
-        if (!serverUri.getScheme().equals("http") && !serverUri.getScheme().equals("https"))
+        if (!serverUri.getScheme().equals("https"))
             return false;
         if (!serverUri.getPath().isEmpty())
             return false;
-        if (serverUri.getPort() != -1)
+        if (serverUri.getPort() <= 0 || serverUri.getPort() > 65535)
             return false;
         return true;
+    }
+
+    private boolean isTrustAllValid(String trustAll) {
+        if (trustAll == null || trustAll.equals("false") || trustAll.equals("true"))
+            return true;
+        return false;
     }
 }

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SecHubClientConfigurationFactory.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SecHubClientConfigurationFactory.java
@@ -1,0 +1,56 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.api.java.demo.cli;
+
+import com.mercedesbenz.sechub.api.java.SecHubClient;
+import com.mercedesbenz.sechub.api.java.demo.cli.exceptions.SecHubClientConfigurationRuntimeException;
+
+public class SecHubClientConfigurationFactory {
+
+    EnvironmentVariableReader environmentVariableReader;
+
+    public SecHubClientConfigurationFactory() {
+        environmentVariableReader = new EnvironmentVariableReader();
+    }
+
+    public SecHubClient create(CommandLineSettings settings) {
+        String serverUri = settings.getServerUri();
+        int serverPort = settings.getServerPort();
+        String userId = settings.getUserId();
+        String apiToken = settings.getApiToken();
+
+        if (serverUri == null) {
+            serverUri = environmentVariableReader.readAsString(EnvironmentVariableConstants.SECHUB_SERVER_URI);
+        }
+        if (serverPort <= 0) {
+            serverPort = environmentVariableReader.readAsInt(EnvironmentVariableConstants.SECHUB_SERVER_PORT);
+        }
+        if (userId == null) {
+            userId = environmentVariableReader.readAsString(EnvironmentVariableConstants.SECHUB_USERID);
+        }
+        if (apiToken == null) {
+            apiToken = environmentVariableReader.readAsString(EnvironmentVariableConstants.SECHUB_APITOKEN);
+        }
+
+        if (serverUri == null) {
+            throw new SecHubClientConfigurationRuntimeException(
+                    "Sechub server's URI is null. Please set the Sechub server's URI as an environment variable (SECHUB_SERVER_URI) or as a program argument (--serverUri).");
+        }
+
+        if (serverPort <= 0) {
+            throw new SecHubClientConfigurationRuntimeException("Sechub server's port is set to " + serverPort
+                    + ". Please set the Sechub server's port as an environment variable (SECHUB_SERVER_PORT) or as a program argument (--serverPort).");
+        }
+
+        if (userId == null) {
+            throw new SecHubClientConfigurationRuntimeException(
+                    "Sechub server's privileged user id is null. Please set the Sechub server's privileged user id as an environment variable (SECHUB_USERID) or as a program argument (--userId).");
+        }
+
+        if (apiToken == null) {
+            throw new SecHubClientConfigurationRuntimeException(
+                    "Privileged user's API token is null. Please set the privileged user's API token as an environment variable (SECHUB_APITOKEN) or as a program argument (--apiToken).");
+        }
+
+        return SecHubClient.create(userId, apiToken, serverUri, serverPort);
+    }
+}

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SecHubClientConfigurationFactory.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SecHubClientConfigurationFactory.java
@@ -1,6 +1,8 @@
 // SPDX-License-Identifier: MIT
 package com.mercedesbenz.sechub.api.java.demo.cli;
 
+import java.net.URI;
+
 import com.mercedesbenz.sechub.api.java.SecHubClient;
 import com.mercedesbenz.sechub.api.java.demo.cli.exceptions.SecHubClientConfigurationRuntimeException;
 
@@ -13,15 +15,19 @@ public class SecHubClientConfigurationFactory {
     }
 
     public SecHubClient create(CommandLineSettings settings) {
-        String serverUri = settings.getServerUri();
+        URI serverUri = null;
+        if (settings.getServerUri() != null)
+            serverUri = URI.create(settings.getServerUri());
+
         int serverPort = settings.getServerPort();
         String userId = settings.getUserId();
         String apiToken = settings.getApiToken();
 
-        if (serverUri == null) {
-            serverUri = environmentVariableReader.readAsString(EnvironmentVariableConstants.SECHUB_SERVER_URI);
+        String envServerUri = environmentVariableReader.readAsString(EnvironmentVariableConstants.SECHUB_SERVER_URI);
+        if (serverUri == null && envServerUri != null) {
+            serverUri = URI.create(envServerUri);
         }
-        if (serverPort <= 0) {
+        if (serverPort <= 0 || serverPort > 65535) {
             serverPort = environmentVariableReader.readAsInt(EnvironmentVariableConstants.SECHUB_SERVER_PORT);
         }
         if (userId == null) {
@@ -31,12 +37,12 @@ public class SecHubClientConfigurationFactory {
             apiToken = environmentVariableReader.readAsString(EnvironmentVariableConstants.SECHUB_APITOKEN);
         }
 
-        if (serverUri == null) {
+        if (!isServerUriValid(serverUri)) {
             throw new SecHubClientConfigurationRuntimeException(
-                    "Sechub server's URI is null. Please set the Sechub server's URI as an environment variable (SECHUB_SERVER_URI) or as a program argument (--serverUri).");
+                    "Sechub server's URI is invalid. Please set the Sechub server's URI as an environment variable (SECHUB_SERVER_URI) or as a program argument (--serverUri).");
         }
 
-        if (serverPort <= 0) {
+        if (serverPort <= 0 || serverPort > 65535) {
             throw new SecHubClientConfigurationRuntimeException("Sechub server's port is set to " + serverPort
                     + ". Please set the Sechub server's port as an environment variable (SECHUB_SERVER_PORT) or as a program argument (--serverPort).");
         }
@@ -51,6 +57,18 @@ public class SecHubClientConfigurationFactory {
                     "Privileged user's API token is null. Please set the privileged user's API token as an environment variable (SECHUB_APITOKEN) or as a program argument (--apiToken).");
         }
 
-        return SecHubClient.create(userId, apiToken, serverUri, serverPort);
+        return SecHubClient.create(userId, apiToken, serverUri.toString(), serverPort);
+    }
+
+    private boolean isServerUriValid(URI serverUri) {
+        if (serverUri == null)
+            return false;
+        if (!serverUri.getScheme().equals("http") && !serverUri.getScheme().equals("https"))
+            return false;
+        if (!serverUri.getPath().isEmpty())
+            return false;
+        if (serverUri.getPort() != -1)
+            return false;
+        return true;
     }
 }

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SechubServerTest.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SechubServerTest.java
@@ -1,0 +1,82 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.api.java.demo.cli;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.mercedesbenz.sechub.api.java.*;
+import com.mercedesbenz.sechub.api.java.auth.Authentication;
+import com.mercedesbenz.sechub.api.java.auth.HttpBasicAuth;
+import com.mercedesbenz.sechub.api.java.demo.cli.exceptions.OpenAPITestToolRuntimeException;
+
+public class SechubServerTest {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SechubServerTest.class);
+    SecHubClient secHubClient = null;
+    ApiClient apiClient = null;
+    StringBuilder fullConnectionURI = new StringBuilder();
+
+    public SechubServerTest(SecHubClient secHubClient) {
+        this.secHubClient = secHubClient;
+        fullConnectionURI.append(secHubClient.getHostUri()).append(":").append(secHubClient.getHostPort());
+    }
+
+    public void runAllTests() {
+        if (!isServerAlive())
+            throw new OpenAPITestToolRuntimeException("Sechub server is not alive.");
+
+        if (!isListOfUsersFetched())
+            throw new OpenAPITestToolRuntimeException("Can not fetch list of users from the Sechub server.");
+    }
+
+    private boolean isServerAlive() {
+        LOG.info("*** Check on whether Sechub server is alive.");
+        apiClient = Configuration.getDefaultApiClient();
+        apiClient.setBasePath(fullConnectionURI.toString());
+
+        AnonymousApi apiInstance = new AnonymousApi(apiClient);
+        try {
+            apiInstance.anonymousCheckAliveGet();
+            LOG.info("*** Sechub server is alive.");
+            return true;
+        } catch (Exception e) {
+            LOG.error("*** Exception when calling AnonymousApi#anonymousCheckAliveGet:");
+            e.printStackTrace();
+        }
+        return false;
+    }
+
+    private boolean isListOfUsersFetched() {
+        LOG.info("*** Trying to fetch list of users.");
+
+        HttpBasicAuth basicAuthentication = new HttpBasicAuth();
+        basicAuthentication.setUsername(secHubClient.getUsername());
+        basicAuthentication.setPassword(secHubClient.getSealedApiToken());
+        Map<String, Authentication> authMap = new HashMap<>();
+        authMap.put("basic", basicAuthentication);
+
+        apiClient = new ApiClient(authMap);
+        apiClient.setBasePath(fullConnectionURI.toString());
+
+        AdminApi apiInstance = new AdminApi(apiClient);
+        try {
+            List<String> usersList = new ArrayList<>();
+            usersList = apiInstance.adminListsAllUsers(basicAuthentication.toString());
+            LOG.info("*** List of users: " + usersList);
+            if (usersList.size() > 0) {
+                LOG.info("*** List of users successfully fetched.");
+                return true;
+            }
+        } catch (ApiException e) {
+            LOG.error("*** Exception when calling AdminApi#adminListsAllUsers:");
+            e.printStackTrace();
+        }
+        return false;
+    }
+
+}

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SechubServerTest.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SechubServerTest.java
@@ -19,11 +19,13 @@ public class SechubServerTest {
     private static final Logger LOG = LoggerFactory.getLogger(SechubServerTest.class);
     SecHubClient secHubClient = null;
     ApiClient apiClient = null;
-    StringBuilder fullConnectionURI = new StringBuilder();
+    String fullConnectionURI = "";
 
     public SechubServerTest(SecHubClient secHubClient) {
         this.secHubClient = secHubClient;
-        fullConnectionURI.append(secHubClient.getHostUri()).append(":").append(secHubClient.getHostPort());
+        StringBuilder sb = new StringBuilder();
+        sb.append(secHubClient.getHostUri()).append(":").append(secHubClient.getHostPort());
+        fullConnectionURI = sb.toString();
     }
 
     public void runAllTests() {
@@ -37,7 +39,7 @@ public class SechubServerTest {
     private boolean isServerAlive() {
         LOG.info("*** Check on whether Sechub server is alive.");
         apiClient = Configuration.getDefaultApiClient();
-        apiClient.setBasePath(fullConnectionURI.toString());
+        apiClient.setBasePath(fullConnectionURI);
 
         AnonymousApi apiInstance = new AnonymousApi(apiClient);
         try {
@@ -61,7 +63,7 @@ public class SechubServerTest {
         authMap.put("basic", basicAuthentication);
 
         apiClient = new ApiClient(authMap);
-        apiClient.setBasePath(fullConnectionURI.toString());
+        apiClient.setBasePath(fullConnectionURI);
 
         AdminApi apiInstance = new AdminApi(apiClient);
         try {

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SechubServerTest.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/SechubServerTest.java
@@ -19,13 +19,9 @@ public class SechubServerTest {
     private static final Logger LOG = LoggerFactory.getLogger(SechubServerTest.class);
     SecHubClient secHubClient = null;
     ApiClient apiClient = null;
-    String fullConnectionURI = "";
 
     public SechubServerTest(SecHubClient secHubClient) {
         this.secHubClient = secHubClient;
-        StringBuilder sb = new StringBuilder();
-        sb.append(secHubClient.getHostUri()).append(":").append(secHubClient.getHostPort());
-        fullConnectionURI = sb.toString();
     }
 
     public void runAllTests() {
@@ -39,7 +35,7 @@ public class SechubServerTest {
     private boolean isServerAlive() {
         LOG.info("*** Check on whether Sechub server is alive.");
         apiClient = Configuration.getDefaultApiClient();
-        apiClient.setBasePath(fullConnectionURI);
+        apiClient.setBasePath(secHubClient.getHostUri().toString());
 
         AnonymousApi apiInstance = new AnonymousApi(apiClient);
         try {
@@ -63,7 +59,7 @@ public class SechubServerTest {
         authMap.put("basic", basicAuthentication);
 
         apiClient = new ApiClient(authMap);
-        apiClient.setBasePath(fullConnectionURI);
+        apiClient.setBasePath(secHubClient.getHostUri().toString());
 
         AdminApi apiInstance = new AdminApi(apiClient);
         try {

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/exceptions/OpenAPITestToolRuntimeException.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/exceptions/OpenAPITestToolRuntimeException.java
@@ -1,0 +1,15 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.api.java.demo.cli.exceptions;
+
+public class OpenAPITestToolRuntimeException extends RuntimeException {
+
+    private static final long serialVersionUID = 1L;
+
+    public OpenAPITestToolRuntimeException(String message) {
+        this(message, null);
+    }
+
+    public OpenAPITestToolRuntimeException(String message, Throwable cause) {
+        super(message, cause);
+    }
+}

--- a/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/exceptions/SecHubClientConfigurationRuntimeException.java
+++ b/sechub-api-java/src/main/java/com/mercedesbenz/sechub/api/java/demo/cli/exceptions/SecHubClientConfigurationRuntimeException.java
@@ -1,0 +1,15 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.api.java.demo.cli.exceptions;
+
+public class SecHubClientConfigurationRuntimeException extends RuntimeException {
+
+    private static final long serialVersionUID = 1L;
+
+    public SecHubClientConfigurationRuntimeException(String message) {
+        this(message, null);
+    }
+
+    public SecHubClientConfigurationRuntimeException(String message, Throwable cause) {
+        super(message, cause);
+    }
+}

